File: frontend/lib/models/negotiation.dart
/// One negotiation — 1:1 with the quotation it's attached to.
class Negotiation {
  final String id;
  final String rfqId;
  final String quotationId;
  final String importerId;
  final String exporterId;
  final String status; // open | accepted | rejected | expired | closed
  final int currentRound;
  final DateTime createdAt;
  final DateTime updatedAt;

  Negotiation({
    required this.id,
    required this.rfqId,
    required this.quotationId,
    required this.importerId,
    required this.exporterId,
    required this.status,
    required this.currentRound,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Negotiation.fromJson(Map<String, dynamic> json) => Negotiation(
        id: json['id'],
        rfqId: json['rfq_id'],
        quotationId: json['quotation_id'],
        importerId: json['importer_id'],
        exporterId: json['exporter_id'],
        status: json['status'],
        currentRound: json['current_round'] ?? 1,
        createdAt: DateTime.parse(json['created_at']),
        updatedAt: DateTime.parse(json['updated_at']),
      );
}

/// One entry in the structured negotiation timeline — the initial quote, every counter
/// offer, and the final accepted offer are all rows of this same shape.
class NegotiationOffer {
  final String id;
  final String negotiationId;
  final int roundNumber;
  final String createdBy; // importer | exporter
  final String createdByUserId;
  final double unitPrice;
  final String currency;
  final double quantity;
  final double? moq;
  final double totalPrice;
  final int? deliveryDays;
  final int? dispatchDays;
  final String? shippingMode;
  final String? incoterm;
  final String? paymentTerms;
  final String? packaging;
  final DateTime? validityDate;
  final String? specialConditions;
  final String? remarks;
  final String status; // pending | accepted | rejected | countered
  final DateTime createdAt;

  NegotiationOffer({
    required this.id,
    required this.negotiationId,
    required this.roundNumber,
    required this.createdBy,
    required this.createdByUserId,
    required this.unitPrice,
    required this.currency,
    required this.quantity,
    this.moq,
    required this.totalPrice,
    this.deliveryDays,
    this.dispatchDays,
    this.shippingMode,
    this.incoterm,
    this.paymentTerms,
    this.packaging,
    this.validityDate,
    this.specialConditions,
    this.remarks,
    required this.status,
    required this.createdAt,
  });

  factory NegotiationOffer.fromJson(Map<String, dynamic> json) => NegotiationOffer(
        id: json['id'],
        negotiationId: json['negotiation_id'],
        roundNumber: json['round_number'],
        createdBy: json['created_by'],
        createdByUserId: json['created_by_user_id'],
        unitPrice: (json['unit_price'] as num).toDouble(),
        currency: json['currency'] ?? 'INR',
        quantity: (json['quantity'] as num).toDouble(),
        moq: (json['moq'] as num?)?.toDouble(),
        totalPrice: (json['total_price'] as num).toDouble(),
        deliveryDays: json['delivery_days'],
        dispatchDays: json['dispatch_days'],
        shippingMode: json['shipping_mode'],
        incoterm: json['incoterm'],
        paymentTerms: json['payment_terms'],
        packaging: json['packaging'],
        validityDate: json['validity_date'] != null ? DateTime.parse(json['validity_date']) : null,
        specialConditions: json['special_conditions'],
        remarks: json['remarks'],
        status: json['status'],
        createdAt: DateTime.parse(json['created_at']),
      );
}

/// Admin negotiations list row — GET /admin/negotiations.
class AdminNegotiationRow {
  final String id;
  final String status;
  final int currentRound;
  final String rfqNumber;
  final String importerName;
  final String exporterName;
  final String productName;
  final int offerCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  AdminNegotiationRow({
    required this.id,
    required this.status,
    required this.currentRound,
    required this.rfqNumber,
    required this.importerName,
    required this.exporterName,
    required this.productName,
    required this.offerCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory AdminNegotiationRow.fromJson(Map<String, dynamic> json) => AdminNegotiationRow(
        id: json['id'],
        status: json['status'],
        currentRound: json['current_round'] ?? 1,
        rfqNumber: json['rfq_number'],
        importerName: json['importer_name'],
        exporterName: json['exporter_name'],
        productName: json['product_name'],
        offerCount: json['offer_count'] ?? 0,
        createdAt: DateTime.parse(json['created_at']),
        updatedAt: DateTime.parse(json['updated_at']),
      );
}

File: backend/internal/models/negotiation.go
package models

import "time"

type NegotiationStatus string

const (
	NegotiationOpen     NegotiationStatus = "open"
	NegotiationAccepted NegotiationStatus = "accepted"
	NegotiationRejected NegotiationStatus = "rejected"
	NegotiationExpired  NegotiationStatus = "expired"
	NegotiationClosed   NegotiationStatus = "closed"
)

type NegotiationOfferStatus string

const (
	OfferPending   NegotiationOfferStatus = "pending"
	OfferAccepted  NegotiationOfferStatus = "accepted"
	OfferRejected  NegotiationOfferStatus = "rejected"
	OfferCountered NegotiationOfferStatus = "countered"
)

type NegotiationParty string

const (
	PartyImporter NegotiationParty = "importer"
	PartyExporter NegotiationParty = "exporter"
)

// Negotiation — one per quotation under negotiation (1:1 with the quotation it's attached to).
type Negotiation struct {
	ID           string            `json:"id" db:"id"`
	RFQID        string            `json:"rfq_id" db:"rfq_id"`
	QuotationID  string            `json:"quotation_id" db:"quotation_id"`
	ImporterID   string            `json:"importer_id" db:"importer_id"`
	ExporterID   string            `json:"exporter_id" db:"exporter_id"`
	Status       NegotiationStatus `json:"status" db:"status"`
	CurrentRound int               `json:"current_round" db:"current_round"`
	CreatedAt    time.Time         `json:"created_at" db:"created_at"`
	UpdatedAt    time.Time         `json:"updated_at" db:"updated_at"`
}

// NegotiationOffer — one entry in the structured timeline (initial quote, each counter
// offer, and the final accepted offer are all rows here).
type NegotiationOffer struct {
	ID                string                 `json:"id" db:"id"`
	NegotiationID     string                 `json:"negotiation_id" db:"negotiation_id"`
	RoundNumber       int                    `json:"round_number" db:"round_number"`
	CreatedBy         NegotiationParty       `json:"created_by" db:"created_by"`
	CreatedByUserID   string                 `json:"created_by_user_id" db:"created_by_user_id"`
	UnitPrice         float64                `json:"unit_price" db:"unit_price"`
	Currency          string                 `json:"currency" db:"currency"`
	Quantity          float64                `json:"quantity" db:"quantity"`
	MOQ               *float64               `json:"moq,omitempty" db:"moq"`
	TotalPrice        float64                `json:"total_price" db:"total_price"`
	DeliveryDays      *int                   `json:"delivery_days,omitempty" db:"delivery_days"`
	DispatchDays      *int                   `json:"dispatch_days,omitempty" db:"dispatch_days"`
	ShippingMode      *string                `json:"shipping_mode,omitempty" db:"shipping_mode"`
	Incoterm          *string                `json:"incoterm,omitempty" db:"incoterm"`
	PaymentTerms      *string                `json:"payment_terms,omitempty" db:"payment_terms"`
	Packaging         *string                `json:"packaging,omitempty" db:"packaging"`
	ValidityDate      *time.Time             `json:"validity_date,omitempty" db:"validity_date"`
	SpecialConditions *string                `json:"special_conditions,omitempty" db:"special_conditions"`
	Remarks           *string                `json:"remarks,omitempty" db:"remarks"`
	Status            NegotiationOfferStatus `json:"status" db:"status"`
	CreatedAt         time.Time              `json:"created_at" db:"created_at"`
}
